File: c-sharp/ChabuCSTest/org/chabu/DataGen.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using StringBuilder = System.Text.StringBuilder;
using RuntimeException = System.SystemException;

namespace org.chabu
{


    public class DataGen
    {

        PseudoRandom gen;
        PseudoRandom exp;
        long genPos = 0;
        long expPos = 0;
        private String name;

        public DataGen(String name, long seed)
        {
            this.name = name;
            gen = new PseudoRandom(seed);
            exp = new PseudoRandom(seed);
        }

        public void getGenBytes(byte[] bytes, int offset, int length)
        {
            gen.nextBytes(bytes, offset, length);
            genPos += length;
        }

        public String getGenBytesString(int numBytes)
        {
            genPos += numBytes;
            return getBytesString(gen, numBytes);
        }

        public void getExpBytes(byte[] bytes, int offset, int length)
        {
            expPos += length;
            exp.nextBytes(bytes, offset, length);
        }

        public String getExpBytesString(int numBytes)
        {
            expPos += numBytes;
            return getBytesString(exp, numBytes);
        }

        private String getBytesString(PseudoRandom rnd, int numBytes)
        {
            if (numBytes == 0)
            {
                return "";
            }
            byte[] bytes = new byte[numBytes];
            rnd.nextBytes(bytes, 0, numBytes);
            StringBuilder sb = new StringBuilder(numBytes * 3 - 1);
            for (int i = 0; i < numBytes; i++)
            {
                if (i > 0) sb.append(" ");
                sb.append(toHexDigit((bytes[i] & 0xF0) >> 4));
                sb.append(toHexDigit((bytes[i] & 0x0F) >> 0));
            }
            return sb.toString();
        }
        private char toHexDigit(int digit)
        {
            if (digit < 10) return (char)('0' + digit);
            return (char)('A' + digit - 10);
        }
        public void ensureSamePosition()
        {
            if (genPos != expPos)
            {
                throw new RuntimeException(String.Format("DataGen {0}, positions not equal: exp:{1} gen:{2}", name, expPos, genPos));
            }
        }
    }
}

File: c/test/src/DataGen.c
#include <stdlib.h>
#include "Common.h"
#include "DataGen.h"
#include "PseudoRandom.h"

static const char* DataGen_GetBytesString(struct PseudoRandom_Data* rnd, int numBytes);
static char toHexDigit(int digit);

void DataGen_Init(struct DataGen_Data* data, const char* name, int64 seed)
{
    data->name = name;
	data->genPos = 0;
	data->expPos = 0;
    PseudoRandom_Init( &data->gen, seed);
    PseudoRandom_Init( &data->exp, seed);
}

void DataGen_GetGenBytes(struct DataGen_Data* data, uint8* bytes, int offset, int length)
{
    PseudoRandom_NextBytes(&data->gen, bytes, offset, length);
    data->genPos += length;
}

//TODO check callers for free'ing result
const char* DataGen_GetGenBytesString(struct DataGen_Data* data, int numBytes)
{
    data->genPos += numBytes;
    return DataGen_GetBytesString(&data->gen, numBytes);
}

void DataGen_GetExpBytes(struct DataGen_Data* data, uint8* bytes, int offset, int length)
{
    data->expPos += length;
    PseudoRandom_NextBytes(&data->exp, bytes, offset, length);
}

//TODO check callers for free'ing result
const char* DataGen_GetExpBytesString(struct DataGen_Data* data, int numBytes)
{
    data->expPos += numBytes;
    return DataGen_GetBytesString( &data->exp, numBytes);
}

static const char* DataGen_GetBytesString(struct PseudoRandom_Data* rnd, int numBytes)
{
	int i, j;
    uint8* bytes;
    char* chars;

	if (numBytes == 0)
    {
        return "";
    }
	
	bytes = (uint8*)calloc(numBytes, 1);
	chars = (char*)calloc(numBytes*3, 1);
    
	PseudoRandom_NextBytes(rnd, bytes, 0, numBytes);
    
	for (i = 0, j = 0; i < numBytes; i++, j += 3)
    {
		chars[j+0] = toHexDigit((bytes[i] & 0xF0) >> 4);
		chars[j+1] = toHexDigit((bytes[i] & 0x0F) >> 0);
        chars[j+2] = ' ';
    }
	free(bytes);
    chars[numBytes*3-1] = '\0';
    return chars;
}
static char toHexDigit(int digit)
{
    if (digit < 10) return (char)('0' + digit);
    return (char)('A' + digit - 10);
}
void DataGen_EnsureSamePosition(struct DataGen_Data* data)
{
    if (data->genPos != data->expPos)
    {
        printf("DataGen %s, positions not equal: exp:%d gen:%d", data->name, data->expPos, data->genPos);
		exit( EXIT_FAILURE);
    }
}
